src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe/moe_gemm.hpp

```diff
@@ -0,0 +1,53 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include "program_node.h"
+#include "registry/implementation_manager.hpp"
+
+using namespace cldnn;  // TODO: Remove once namespaces are aligned
+
+namespace ov::intel_gpu::ocl {
+
+struct MoEGemm : public ImplementationManager {
+    OV_GPU_PRIMITIVE_IMPL("ocl::moe_gemm")
+
+    explicit MoEGemm(shape_types shape_type, ValidateFunc vf = nullptr) : ImplementationManager(impl_types::ocl, shape_type, std::move(vf)) {}
+
+    [[nodiscard]] std::unique_ptr<primitive_impl> create_impl(const program_node& node, const RuntimeParams& params) const override;
+
+    [[nodiscard]] bool validate_impl(const program_node& node) const override {
+        const auto& device_info = node.get_program().get_engine().get_device_info();
+        // TODO: only micro kernel is available now for moe_gemm
+        if (!device_info.supports_immad)
+            return false;
+
+        static const std::vector<format> supported_fmts = {
+            format::bfyx,
+        };
+
+        static const std::vector<ov::element::Type_t> supported_types = {
+            ov::element::f32,
+            ov::element::f16,
+            ov::element::u4,
+            ov::element::i4,
+            ov::element::i8,
+            ov::element::u8,
+            ov::element::i32,
+        };
+
+        for (const auto& input_layout : node.get_input_layouts()) {
+            if (!one_of(input_layout.format, supported_fmts) || !one_of(input_layout.data_type, supported_types)) {
+                return false;
+            }
+        }
```

Comment:
* does micro gemm support fp32 as well? (dpas doesn't support fp32 as far as I remember..)
* nit: I guess i4/u4 is not supported from activation

Reply:
Updated this part to handle each input